### ISSUE TITLE
Pubrec error 145

### ISSUE THREAD

**Issue by nitrogene** [eclipse/paho.mqtt.c#1279]
Hello,

We are using *paho-mqtt-cpp/1.2.0* and its dependency *paho-mqtt-c/1.3.9* in our C++ project to publish / subscribe to an emqx broker *4.3.9* embedded in a docker container. We are using the robot framework python library  to test our software. From time to time, when executing a given test, our software failed to publish a message.
By enabling traces as explained [here](https://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/tracing.html), I was able to get more information:

```log
20221018 081318.689 6 XXXX<- PUBREC msgid: 29
20221018 081318.689 Pubrec error 145 received for client XXXX 29, not sending PUBREL
20221018 081318.689 (2373396224)    (3)> MQTTPersistence_remove:537
20221018 081318.689 (2373396224)    (3)< MQTTPersistence_remove:595 (0)
20221018 081318.689 (2373396224)    (3)> MQTTProtocol_removePublication:289
20221018 081318.689 (2373396224)    (3)< MQTTProtocol_removePublication:298
20221018 081318.689 (2373396224)    (3)> MQTTProperties_free:401
20221018 081318.689 (2373396224)    (3)< MQTTProperties_free:424
20221018 081318.689 (2373396224)    (3)> MQTTProperties_free:401
20221018 081318.689 (2373396224)    (3)< MQTTProperties_free:424
20221018 081318.689 (2373396224)   (2)< MQTTProtocol_handlePubrecs:515 (0)
20221018 081318.689 Calling publish failure for client XXXX
```
I am not sure how to investigate. Is an *emqx* issue ? Should we surround each publish with a try/catch and retry the publish in the catch section ? 

Best regards,

Jean-Pierre

**Comment by icraggs** [comment]
It looks like you are using MQTT 5.0?

The client library is reporting it has received a reason code 145 from the broker, PACKET_IDENTIFIER_IN_USE.

A library trace at the PROTOCOL level might provide more insight.

**Comment by Lenormju** [comment]
I'm a coworker of @nitrogene, I encountered the problem again and ended up here.

We are indeed using v5. Since the initial report, we have added more debug information, both logging which application messages have been correctly received by other clients, and the protocol exchanges with the broker.  
We can confirm that the other clients never received the message that was failing to send with the `"Packet Identifier in use"` error (`PACKET_IDENTIFIER_IN_USE`).

This is further detailed in the protocol diff between a non-working and a working case :  
![diff_protocol](https://user-images.githubusercontent.com/13200639/231480081-37cf951c-5f56-4bfb-a014-34ef724df919.png)
(the name has been obfuscated)

We can see that the first PUBLISH `qos=2` received a PUBREC, but there was no PUBREL+PUBCOMP following, we assume because the MQTT client raised the PACKET_IDENTIFIER_IN_USE error received from the broker.
As we re-use the same client name between test runs, and because we connect with `clean=0` (not clean, i.e. reusing the previous session state), we suspect there is a message stuck (for whatever reason) in the broker, preventing the client to push a new message witj this `msgid`.
The second PUBLISH worked fine, so it looks like a stuck message. Restarting our broker (EMQX) seems to have solved the problem again.

In conclusion, we do not think there is any problem with the Paho library, we will look at the broker instead. Thanks anyway for your indications.

**Comment by icraggs** [comment]
Thank you for the information. I will close this issue. If you want to discuss further in the future, you can reopen this issue.

**Comment by YraganTron** [comment]
@Lenormju Hi! I have same problem. I have paho.mqtt.c client and emqx broker. And my QoS 2 Message Receive Queue sometimes full for no obvious reason. I get 145 code and 0 code error for publish.
Maybe you have discovered the reason for this behavior in the broker?



```
Can't publish the message (id 475)to the ['events/diagnostics'] topic(s). Reason code 145.
Can't publish the message (id 0)to the ['b/events/diagnostics'] topic(s). Reason code 0.
```

**Comment by nitrogene** [comment]
I'm sorry, we didn't have the time to fully investigate this issue.
